jetson/percep/pcl.cpp

```diff
@@ -160,25 +160,60 @@ void PCL::FindInterestPoints(std::vector<pcl::PointIndices> &cluster_indices,
         for (auto index : cluster_indices[i].indices)
         {
             auto curr_point = pt_cloud_ptr->points[index];
             
-            //Upper Left
-            if(curr_point.x < pt_cloud_ptr->points[curr_cluster->at(0)].x && curr_point.y > pt_cloud_ptr->points[curr_cluster->at(0)].y)
+            if(curr_point.x < pt_cloud_ptr->points[curr_cluster->at(0)].x){
                 curr_cluster->at(0) = index;
-            //Upper Right
-            else if(curr_point.x > pt_cloud_ptr->points[curr_cluster->at(1)].x && curr_point.y > pt_cloud_ptr->points[curr_cluster->at(1)].y)
+            }
+            if(curr_point.x > pt_cloud_ptr->points[curr_cluster->at(1)].x){
                 curr_cluster->at(1) = index;
-            //Low Left
-            else if(curr_point.x < pt_cloud_ptr->points[curr_cluster->at(2)].x && curr_point.y < pt_cloud_ptr->points[curr_cluster->at(2)].y)
+            }
+            if(curr_point.y < pt_cloud_ptr->points[curr_cluster->at(2)].y){
                 curr_cluster->at(2) = index;
-            //Low Right
-            else if(curr_point.x > pt_cloud_ptr->points[curr_cluster->at(3)].x && curr_point.y < pt_cloud_ptr->points[curr_cluster->at(3)].y)
+            }
+            if(curr_point.y > pt_cloud_ptr->points[curr_cluster->at(3)].y){
                 curr_cluster->at(3) = index;
+            }
+            if(curr_point.z < pt_cloud_ptr->points[curr_cluster->at(4)].z){
+                curr_cluster->at(4) = index;
+            }
+            if(curr_point.z > pt_cloud_ptr->points[curr_cluster->at(5)].z){
+                curr_cluster->at(5) = index;
+            }
+        }
+
+        //Calulates the width of the obstacle based the difference between the leftmost and rightmost interst point
+        double width = std::abs(pt_cloud_ptr->points[curr_cluster->at(1)].x - pt_cloud_ptr->points[curr_cluster->at(0)].x);
+        //Calculates the number of rover widths that fit within the obstacle. The x10 multiplier adds more width increments.
+        int rovers = ((int) width/ROVER_W_MM) * 10;
+        //Creates a vector called points which holds the width increments we want the new interest points to be close to.
+        //Example: if rovers = 4, then the points vector would contain:  0.025, 0.05, 0.075, 0.1, 0.125, ..., 0.975
+        //The size of this vector will be the amount of new interest points we are adding to this obstacle.
+        std::vector<double> points;
+        for(double i = ((double) 1/rovers) ; i < 0.99999; i += ((double) 1/rovers)){
+            points.push_back(((double)i * width));
+        }
+
+        //Sets the new interest points equal to the leftmost point.
+        for(size_t i = 0; i < points.size(); ++i){
+            curr_cluster->at(9 + i) = curr_cluster->at(0);
         }
+
+        //Finds the new interest points based on width increments.
+        for (auto index : cluster_indices[i].indices){
```

Comment:
This to me seems very expensive especially if we have a large number in the points vector. Since we know the left and rightmost points are you able to use the x value of the current point, calculate the percentile that the current point would fall under, and then compare that x value to the one of the point that is currently representing that percentile? I think a good data structure to use for this would be std::map. If rovers = 40 and if your x value falls between leftmost + 0.025*width and leftmost+0.05*width you could have a function that would generate an index of 1 and could then lookup the current value you have stored at index 1 in your map?

Reply:
Could you also compare the time that this function takes now compared to the time it takes the code that is currently in main and post times in a comment? You can run it against test1 or test2 image files to find this out

Reply:
Implemented the map